drivers: pwm: add STM32 LPTIM PWM driver

tpambor opened the pull request:
This PR introduces a PWM driver for the STM32 low-power timer (LPTIM) peripheral on STM32 MCUs.

This driver is loosely based on the PWM driver for standard timers on STM32. For maintainability and readability, I split this into a new driver instead of integrating with the existing STM PWM driver, as the supported features differ and LPTIMs use a different low-level API `LL_LPTIM_*`.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/57316.

gautierg-st commented on drivers/pwm/pwm_stm32_lptim.c:
```diff
@@ -0,0 +1,249 @@
+		return LL_LPTIM_OUTPUT_POLARITY_INVERSE;
+	}
+
+	return LL_LPTIM_OUTPUT_POLARITY_REGULAR;
```
```suggestion
	return (flags & PWM_POLARITY_MASK) == PWM_POLARITY_NORMAL ?
	        LL_LPTIM_OUTPUT_POLARITY_INVERSE : LL_LPTIM_OUTPUT_POLARITY_REGULAR;
```

gautierg-st reviewed (changes requested):
Thanks for this. Driver side looks ok to me.
As I said in one comment, a `pwm` child node should be added in the `lptim` nodes of the dtsi, like it is done for the `timers` nodes. This will require some minor changes to the driver and the overlay.
And make that 3 commits:
- Driver
- Dtsi update
- Tests/samples

etienne-lms commented on drivers/pwm/pwm_stm32_lptim.c:
```diff
@@ -0,0 +1,249 @@
+
+	/* Initialize timer */
+	LL_LPTIM_SetClockSource(timer, LL_LPTIM_CLK_SOURCE_INTERNAL);
+	LL_LPTIM_SetPrescaler(timer, (__CLZ(__RBIT(cfg->prescaler)) << LPTIM_CFGR_PRESC_Pos));
```
Nitpicking: could remove an inner parentheses pair:
```suggestion
	LL_LPTIM_SetPrescaler(timer, __CLZ(__RBIT(cfg->prescaler)) << LPTIM_CFGR_PRESC_Pos);

tpambor commented on drivers/pwm/pwm_stm32_lptim.c:
```diff
@@ -0,0 +1,249 @@
+	return 0;
+}
+
+#define PWM(index) DT_DRV_INST(index)
```
I generally agree, but this would require major refactoring and introduce breaking changes. The current top-level `st,stm32-lptim` binding is tailored specifically for using LPTIM as a system tick source. Therefore, I see the following steps:
1. Introduce a new top-level binding: `st,stm32-lptimers`, essentially based on `st,stm32-lptim` but without the tick-source-specific `st,timeout` property.
2. Add a new binding for tick source functionality and include it as a subnode within LPTIM instances.
3. Update the current LPTIM tick source driver to align with the new structure.
4. Update samples and tests to reflect these changes.
5. Update the binding of this PWM driver so that it can be added as a subnode to LPTIM instances that support PWM.
6. Update this PWM driver

All users using LPTIM as systick source would then need to update their device tree to adapt to these changes.

FRASTM replied:
> > Introduce a new top-level binding: st,stm32-lptimers, essentially based on st,stm32-lptim but without the tick-source-specific st,timeout property.
> 
> Why is that? I must be missing something, the property is not required, so just don't use it in the PWM driver.
> 
> [zephyr/dts/bindings/timer/st,stm32-lptim.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/72360f821a610b64fda3bd300b6ca4577f6abe8c/dts/bindings/timer/st%2Cstm32-lptim.yaml#L48)
> 
> Line 48 in [72360f8](/zephyrproject-rtos/zephyr/commit/72360f821a610b64fda3bd300b6ca4577f6abe8c)
> 
>  st,timeout:

As @JarmouniA says, the `st,timeout` is not a "required" property, 
Then [‎dts/bindings/pwm/st,stm32-lptim-pwm.yaml](https://github.com/zephyrproject-rtos/zephyr/pull/101205/changes/6caabeef934bbe77a112e8af85c22ccbae713772#diff-e5805b235f97413bbd49bb0ed490942b9fc23fb646b2da7d664090587bb2a6be) is not needed, a compatible: "st,stm32-pwm"  is fine

tpambor committed:
drivers: pwm: add STM32 LPTIM PWM driver

Add a PWM driver for the STM32 low-power timer (LPTIM) peripheral
on STM32 MCUs.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

tpambor committed:
samples: drivers: led: pwm: Add support for stm32h573i_dk

Add overlay and update sample.yaml to include stm32h573i_dk board.

This board can utilize LPTIM5 for PWM generation to drive the
red (LD3) LED.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

FRASTM commented:
@tpambor do you plan to continue working on that PR ?

tpambor commented:
> @tpambor do you plan to continue working on that PR ?

I am still interested but I currently have no plans in short term to work on this.
